MBS-11977: Fix targeting ENTITY for URL relationship error

yvanzo opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Server.
    We appreciate your time and interest in helping our project!

    Please use this template to help us review your change.

    Depending on your change, some sections may be unneeded, just remove these.
    For example, small pull requests usually don’t need the section “Action”.

    Remember that the more helpful info your pull request includes,
    the easier it is for us to understand and review your changes.

    Ensure that you’ve read through and followed the Contributing Guidelines, at
    https://github.com/metabrainz/musicbrainz-server/blob/master/CONTRIBUTING.md
-->

# Problem
<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

MBS-11977: Mismatched entity type in external links editor doesn't trigger the appropriate error message

(see the ticket for details)

# Solution
<!--
    Talk about technical details, considerations, or other interesting points.
-->

Fix targeting `ENTITY` for URL relationship error.

It also required adding a generic error message for `ENTITY` error target (which was associated to custom error message only so far).

# Notes

Further improvement would be to specify which entity type should be linked to the entered URL, and which URL pattern would be suitable for the current entity, as proposed in “Implementation notes” of the ticket. Actually, it is probably worth a separate ticket.

# Checklist for author

* [ ] Rebase on `master` if #2263 is merged first

reosarevok commented on root/static/scripts/edit/externalLinks.js:
```diff
@@ -615,6 +616,14 @@ export class ExternalLinksEditor
+        if (error.target ===
+          URLCleanup.ERROR_TARGETS.ENTITY) {
+          error.message = exp.l(
+            `This URL is not allowed for this type of entity:
+             {entity_type_name}.`,
```
Maybe "This URL is not allowed for “{entity_type_name}” entities" is more straightforward? (while still being translatable).

yvanzo replied:
Applied with https://github.com/metabrainz/musicbrainz-server/pull/2280/commits/277ad5922874f54f4f9077b13d8ac195aaec2764 commit.

reosarevok commented:
@brainzbot, retest this please

reosarevok reviewed (approved):
Seems sensible, in general. I do wonder whether this error should only be shown if the link is valid for *other* entity type, but I expect in most cases this is going to be fine.

yvanzo commented:
Yes it should “be fine in most cases” and `URL` should probably be the target instead “if the link is _not_ valid for any other entity type”.

mwiencek commented on root/static/scripts/edit/externalLinks.js:
```diff
@@ -615,6 +616,13 @@ export class ExternalLinksEditor
           error.message = l(`This URL is not allowed 
                     for the selected link type.`);
         }
+        if (error.target ===
+          URLCleanup.ERROR_TARGETS.ENTITY) {
```
```suggestion
        if (error.target === URLCleanup.ERROR_TARGETS.ENTITY) {
```

mwiencek commented on root/static/scripts/edit/externalLinks.js:
```diff
@@ -615,6 +616,13 @@ export class ExternalLinksEditor
+        if (error.target ===
+          URLCleanup.ERROR_TARGETS.ENTITY) {
+          error.message = exp.l(
+            `This URL is not allowed for “{entity_type_name}” entities.`,
+            {entity_type_name: formatEntityTypeName(checker.entityType)},
```
I think a separate string for each entity type (like we do for SEARCH_PLACEHOLDERS, "Add a new x," "Add another x," etc. would be nice here; saying `“artist” entities` instead of `artists` just looks and sounds awkward. Like `Yesterday I watched a movie of the genre “horror”` or something. :laughing:

reosarevok replied:
That is fair, it shouldn't be a crazy amount of strings and they should be easy to translate anyway :)

yvanzo replied:
Thanks, I was just too lazy here. Split in the commit <https://github.com/metabrainz/musicbrainz-server/pull/2280/commits/cb27c59e6d85b62f1ef77aa4129f637c6f9fd1e6>.

mwiencek reviewed (approved):
Skimming the changes, they look reasonable, though I can't test all the changes here.  I guess the automated tests don't check the expected error target yet?

yvanzo commented:
> Skimming the changes, they look reasonable, though I can't test all the changes here.

I didn’t test all of them either. I fixed the most common anti-patterns only. Some of the functions `validate` will require larger refactoring, but that will be more closely double-checked when improving error messages.

> I guess the automated tests don't check the expected error target yet?

Right, most of the new summer features are roughly tested using Selenium only. The file `root/static/scripts/tests/Control/URLCleanup.js` should be updated with further tests. Just added  MBS-11995. (See also MBS-11980 for rel types restriction.)
